### PR TITLE
Add realized armor for yin and yang

### PR DESCRIPTION
**Pull request by CrabbytheCrab**
## About The Pull Request

Adds realized armors both both yin and yang called harmony of duality and duality of harmony respectively.

Harmony of duality has  9/4/9/4 and has an ability called Anarchy. Anarchy is a targetable realized ability with a minute cooldown that lets you fire a laser dealing 175 black damage and if you hit another player(or yourself) with it they'll gain a + 20 justice buff for a minute and a half. If someone gets hir by Anarchy and is wearing Duality of harmony they'll get a buff that decreases red and pale black taken by 25% and increases their stats by 10 for 90 seconds.

Duality of harmony has 9/4/9/4 and has an ability called Tranquility. Tranquility is a charge up ability with a minute cool down that lets you explode like how yang does on death dealing 300 white damage and healing people caught in its blast for 120 hp and sp. If someone gets healed by Tranquility and is wearing Harmony of duality they'll get a buff that decreases white and pale damage taken by 25% and increases their stats by 10 for 90 seconds.

Duality of harmony and Harmony of duality sound broken as fuck(they probably need to be toned down) but here's a catch, you need to have both yin and yang to be able to realize them and having both yin and yang are quite risky already.

## Why It's Good For The Game

Adding realized armors for both yin and yang that requires having both makes having the 2 more worthwhile and makes the dragon of harmony show up more often

## Changelog
:cl:
add: duality of harmony, harmony of duality, and their abilities and buffs
imageadd: added some icons and images
code: adds a check for the realizer for having both yin and yang
/:cl:

### CODE REVIEW

File: code/modules/spells/ability_types/realized.dm [hunk 1]
```diff
@@ -789,3 +790,87 @@
...
+
+/datum/status_effect/duality_yang
+	id = "EGO_YANG"
+	status_type = STATUS_EFFECT_UNIQUE
+	alert_type = /atom/movable/screen/alert/status_effect/duality_yang
+	duration = 90 SECONDS
+
+/atom/movable/screen/alert/status_effect/duality_yang
+	name = "Duality of harmony"
+	desc = "Decrese white and pale damage taken by 25%. \
```
Comment (Coxswain-Navigator) [inline code/modules/spells/ability_types/realized.dm:854]: typo

File: code/modules/spells/ability_types/realized_aimed.dm [hunk 2]
```diff
@@ -411,7 +411,128 @@
...
+		sleep(1)
+		new /obj/effect/temp_visual/revenant/cracks/yinfriend(OT)
+	for(var/obj/effect/FX in spawned_effects)
+		qdel(FX)
+	qdel(beam)
+
+/obj/effect/temp_visual/revenant/cracks/yinfriend
+	icon_state = "yincracks"
+	duration = 9
+	var/damage = 1000  // Amount of black damage dealt to enemies from the laser.
```
Comment (Coxswain-Navigator) [inline code/modules/spells/ability_types/realized_aimed.dm:456]: You don't think 1000 damage in an AOE on a 60 second cooldown is too much?

**Comment by Kitsunemitsu** [comment]
Don't notate in decimals.
Notation on PRs should be #/#/#/#
Such as 5/6/6/5 or 6/6/6/6

**Comment by Kitsunemitsu** [comment]
Also what's the point of all these buffs........

**Comment by CrabbytheCrab** [comment]
> Also what's the point of all these buffs........
> 
> Applying DNM until author can explain the reason of their buffs and nerfs, seemingly at random

**Comment by CrabbytheCrab** [comment]
those balancing changes were due to me asking about which of the realized armors a pr made before needed to be buffed or nerfed and quenched with blood was brought up during one. im willing to undo sakura bloom's, eyes of god, and fallen colors changes since those we're by decision fully

**Comment by CrabbytheCrab** [comment]
did the split

**Comment by Kitsunemitsu** [comment]
Balancing changes removed.

File: code/modules/clothing/suits/ego_gear/realized.dm [hunk 3]
```diff
@@ -283,6 +283,21 @@
 	armor = list(RED_DAMAGE = 70, WHITE_DAMAGE = 80, BLACK_DAMAGE = 70, PALE_DAMAGE = 40)
 	realized_ability = /obj/effect/proc_holder/ability/shrimp
 
+/obj/item/clothing/suit/armor/ego_gear/realization/duality_yang
+	name = "duality of harmony"
+	desc = "When good an evil meet discord and assonance will be quelled."
```
Comment (LanceSmites328) [inline code/modules/clothing/suits/ego_gear/realized.dm:288]:
```suggestion
	desc = "When good and evil meet discord and assonance will be quelled."
```

File: code/modules/mob/living/simple_animal/abnormality/_tools/zayin/realizer.dm [hunk 4]
```diff
@@ -52,15 +54,29 @@
...
+	var yang_armor = /obj/item/clothing/suit/armor/ego_gear/waw/assonance
+	var yin_armor = /obj/item/clothing/suit/armor/ego_gear/waw/discord
 	if(!ishuman(user))
 		return
 
 	if(!(I.type in output))
 		to_chat(user, "<span class='warning'>The true potential of [I] cannot be realized.</span>")
 		return
 
+	if(I.type == yang_armor && !YinYangCheck() || I.type == yin_armor && !YinYangCheck())
```
Comment (LanceSmites328) [inline code/modules/mob/living/simple_animal/abnormality/_tools/zayin/realizer.dm:76]:
```suggestion
	if((istype(I, /obj/item/clothing/suit/armor/ego_gear/waw/discord) || istype(I, /obj/item/clothing/suit/armor/ego_gear/waw/assonance) && !YinYangCheck())
```

File: code/modules/mob/living/simple_animal/abnormality/_tools/zayin/realizer.dm [hunk 5]
```diff
@@ -52,15 +54,29 @@
...
+		if(AD.abno_path == /mob/living/simple_animal/hostile/abnormality/yang)
+			for(var/datum/abnormality/AD2 in SSlobotomy_corp.all_abnormality_datums)
+				if(AD2.abno_path == /mob/living/simple_animal/hostile/abnormality/yin)
+					return TRUE
+	return FALSE
+
 /obj/structure/toolabnormality/realization/attackby(obj/item/I, mob/living/carbon/human/user)
 	. = ..()
+	var yang_armor = /obj/item/clothing/suit/armor/ego_gear/waw/assonance
+	var yin_armor = /obj/item/clothing/suit/armor/ego_gear/waw/discord
```
Comment (LanceSmites328) [inline code/modules/mob/living/simple_animal/abnormality/_tools/zayin/realizer.dm:68]:
```suggestion
```

File: code/modules/spells/ability_types/realized.dm [hunk 6]
```diff
@@ -789,3 +789,88 @@
...
+	desc = "An ability that does massive white damage in an area and heals people's health and sanity. \
+	Healing someone that's wearing harmony of duality will grant them huge buffs to their defenses and stats."
+	action_icon_state = "yangform0"
+	base_icon_state = "yangform"
+	cooldown_time = 60 SECONDS
+
+	var/damage_amount = 300 // Amount of explosion damage
+	var/explosion_range = 15
+
+/obj/effect/proc_holder/ability/tranquility/Perform(target, mob/user)
```
Comment (LanceSmites328) [inline code/modules/spells/ability_types/realized.dm:805]:
```suggestion
/obj/effect/proc_holder/ability/tranquility/Perform(target, mob/living/carbon/human/user)
```

File: code/modules/spells/ability_types/realized.dm [hunk 7]
```diff
@@ -789,3 +789,88 @@
...
+					continue
+				if(L.is_working) //no work heal :(
+					continue
+				L.adjustBruteLoss(-120)
+				L.adjustSanityLoss(-120)
+				var/obj/item/clothing/suit/armor/ego_gear/realization/duality_yin/Z = L.get_item_by_slot(ITEM_SLOT_OCLOTHING)
+				if(istype(Z))
+					H.apply_status_effect(/datum/status_effect/duality_yang)
+					new /obj/effect/temp_visual/healing(get_turf(L))
+			all_turfs -= T
```
Comment (LanceSmites328) [inline code/modules/spells/ability_types/realized.dm:840]:
```suggestion
	for(var/i = 0 to explosion_range)
		for(var/turf/T in all_turfs)
			if(get_dist(user, T) > i)
				continue
			new /obj/effect/temp_visual/dir_setting/speedbike_trail(T)
			user.HurtInTurf(damage_amount, list(), WHITE_DAMAGE)
			for(var/mob/living/carbon/human/L in T)
				if(!user.faction_check_mob(L, FALSE))
					continue
				if(L.stat == DEAD)
					continue
				if(L.is_working) //no work heal :(
					continue
				L.adjustBruteLoss(-120)
				L.adjustSanityLoss(-120)
				if(istype(L.get_item_by_slot(ITEM_SLOT_OCLOTHING), /obj/item/clothing/suit/armor/ego_gear/realization/duality_yin))
					H.apply_status_effect(/datum/status_effect/duality_yang)
					new /obj/effect/temp_visual/healing(get_turf(L))
			all_turfs -= T
```
I will also note that the healing VFX only applies to those who have the opposing suit on at the moment.

File: code/modules/spells/ability_types/realized_aimed.dm [hunk 8]
```diff
@@ -415,3 +415,124 @@
...
+			L.apply_damage(damage, BLACK_DAMAGE, null, L.run_armor_check(null, BLACK_DAMAGE), spread_damage = TRUE)
+		for(var/mob/living/carbon/human/L in T)
+			if(!faction_check(L.faction, src.faction))
+				continue
+			if(L.status_flags & GODMODE)
+				continue
+			if(L.stat == DEAD)
+				continue
+			if(L.is_working) //no work heal :(
+				continue
```
Comment (LanceSmites328) [inline code/modules/spells/ability_types/realized_aimed.dm:475]:
```suggestion
```
None of these checks are particularly relevant to any scenario that this buff would be disruptive to.

File: code/modules/spells/ability_types/realized_aimed.dm [hunk 9]
```diff
@@ -415,3 +415,124 @@
...
+			L.apply_status_effect(/datum/status_effect/yinboost)
+			var/obj/item/clothing/suit/armor/ego_gear/realization/duality_yang/Z = L.get_item_by_slot(ITEM_SLOT_OCLOTHING)
+			if(istype(Z))
```
Comment (LanceSmites328) [inline code/modules/spells/ability_types/realized_aimed.dm:478]:
```suggestion
			if(istype(L.get_item_by_slot(ITEM_SLOT_OCLOTHING), /obj/item/clothing/suit/armor/ego_gear/realization/duality_yang))
```

File: code/modules/spells/ability_types/realized_aimed.dm [hunk 10]
```diff
@@ -415,3 +415,124 @@
...
+/obj/effect/proc_holder/ability/aimed/yin_laser
+	name = "Anarchy"
+	desc = "An ability that summons a devastating laser. \
+		If another player gets hit by the laser they get +20 justice for a period of time. \
+		If the person that gets hit is wearing duality of harmony, they will get huge buffs to their defenses and stats."
+	action_icon_state = "yinform0"
+	base_icon_state = "yinform"
+	cooldown_time = 60 SECONDS
+	var/laser_range = 20
+	var/list/spawned_effects = list()
```
Comment (LanceSmites328) [inline code/modules/spells/ability_types/realized_aimed.dm:428]:
```suggestion
```
You don't do anything with this list, nothing's added to it.

File: code/modules/spells/ability_types/realized_aimed.dm [hunk 11]
```diff
@@ -415,3 +415,124 @@
...
+	var/datum/beam/beam  = my_turf.Beam(target_turf, "volt_ray")
+	for(var/turf/open/OT in to_hit)
+		if(!istype(OT) || OT.density)
+			break
+		beam.target = OT
+		beam.redrawing()
+		sleep(1)
+		new /obj/effect/temp_visual/revenant/cracks/yinfriend(OT)
+	for(var/obj/effect/FX in spawned_effects)
+		qdel(FX)
```
Comment (LanceSmites328) [inline code/modules/spells/ability_types/realized_aimed.dm:450]:
```suggestion
```

File: code/modules/spells/ability_types/realized_aimed.dm [hunk 12]
```diff
@@ -415,3 +415,124 @@
...
+
+/obj/effect/proc_holder/ability/aimed/yin_laser/Perform(target, user)
+	var/turf/t_turf = get_turf(target)
+	INVOKE_ASYNC(src, .proc/Cast, t_turf, user)
+	return ..()
+
+/obj/effect/proc_holder/ability/aimed/yin_laser/proc/Cast(turf/target, mob/user)
+	user.face_atom(target)
+	var/mob/living/carbon/human/H = user
+	var/turf/my_turf = get_turf(H)
```
Comment (LanceSmites328) [inline code/modules/spells/ability_types/realized_aimed.dm:438]:
```suggestion
/obj/effect/proc_holder/ability/aimed/yin_laser/proc/Cast(turf/target, mob/living/carbon/human/user)
	user.face_atom(target)
	var/turf/my_turf = get_turf(user)
```

**Review by LanceSmites328 — CHANGES_REQUESTED** [review]

**Review by LanceSmites328 — APPROVED** [review]
If the numbers on these pieces of EGO are a problem in the future, we can change them. They are currently implemented as is as managing both Yin and Yang is risky as hell and both getting out can spell disaster for a facility.

**Pull request merged**